tests/src/test_helpers/ssl_helpers.c

```diff
@@ -0,0 +1,2536 @@
+int mbedtls_test_message_socket_setup(
+        mbedtls_test_ssl_message_queue* queue_input,
+        mbedtls_test_ssl_message_queue* queue_output,
+        size_t queue_capacity, mbedtls_test_mock_socket* socket,
+        mbedtls_test_message_socket_context* ctx )
+{
+    int ret = mbedtls_test_ssl_message_queue_setup( queue_input,
+                                                    queue_capacity );
+    if( ret != 0 )
+        return ret;
+    ctx->queue_input = queue_input;
+    ctx->queue_output = queue_output;
+    ctx->socket = socket;
+    mbedtls_mock_socket_init( socket );
+
+    return 0;
+}
+
+/*
+ * Close a given message socket context, along with the socket itself. Free the
+ * memory allocated by the input queue.
+ */
+void mbedtls_test_message_socket_close(
+        mbedtls_test_message_socket_context* ctx )
+{
+    if( ctx == NULL )
+        return;
+
+    mbedtls_test_ssl_message_queue_free( ctx->queue_input );
+    mbedtls_test_mock_socket_close( ctx->socket );
+    memset( ctx, 0, sizeof( *ctx ) );
+}
+
+/*
+ * Send one message through a given message socket context.
+ *
+ * \retval  \p len, if everything succeeds.
+ * \retval  MBEDTLS_TEST_ERROR_CONTEXT_ERROR, if any of the needed context
+ *          elements or the context itself is null.
+ * \retval  MBEDTLS_TEST_ERROR_SEND_FAILED if
+ *          mbedtls_test_mock_tcp_send_b failed.
+ * \retval  MBEDTLS_ERR_SSL_WANT_WRITE, if the output queue is full.
+ *
+ * This function will also return any error from
+ * mbedtls_test_ssl_message_queue_push_info.
+ */
+int mbedtls_test_mock_tcp_send_msg( void *ctx, const unsigned char *buf,
+                                    size_t len )
+{
+    mbedtls_test_ssl_message_queue* queue;
+    mbedtls_test_mock_socket* socket;
+    mbedtls_test_message_socket_context *context =
+        (mbedtls_test_message_socket_context*) ctx;
+
+    if( context == NULL || context->socket == NULL
+        || context->queue_output == NULL )
+    {
+        return MBEDTLS_TEST_ERROR_CONTEXT_ERROR;
+    }
+
+    queue = context->queue_output;
+    socket = context->socket;
+
+    if( queue->num >= queue->capacity )
+        return MBEDTLS_ERR_SSL_WANT_WRITE;
+
+    if( mbedtls_test_mock_tcp_send_b( socket, buf, len ) != (int) len )
+        return MBEDTLS_TEST_ERROR_SEND_FAILED;
+
+    return mbedtls_test_ssl_message_queue_push_info( queue, len );
+}
+
+/*
+ * Receive one message from a given message socket context and return message
+ * length or an error.
+ *
+ * \retval  message length, if everything succeeds.
+ * \retval  MBEDTLS_TEST_ERROR_CONTEXT_ERROR, if any of the needed context
+ *          elements or the context itself is null.
+ * \retval  MBEDTLS_TEST_ERROR_RECV_FAILED if
+ *          mbedtls_test_mock_tcp_recv_b failed.
+ *
+ * This function will also return any error other than
+ * MBEDTLS_TEST_ERROR_MESSAGE_TRUNCATED from
+ * mbedtls_test_message_queue_peek_info.
+ */
+int mbedtls_test_mock_tcp_recv_msg( void *ctx, unsigned char *buf,
```

Comment:
```suggestion
int mbedtls_test_mock_tcp_recv_msg( void *ctx,
                                                             unsigned char *buf,
```
Prefer to put buf and buf_len in the same line.